Minor improvements

KSVdvlpr opened the pull request.

KSVdvlpr committed:
Second screen MVP

KSVdvlpr committed:
Second screen MVP
- with universal method onCurrencyConverted

KSVdvlpr committed:
- naming
- swipe to refresh
- dialog

KSVdvlpr committed:
Доюавил:
- подключил библиотеку  корутин
- исправил поведение конвертера на втором экране
- добавил диалог на первый и второй экран при отсутствии интернета (с возможнстью прехода в настройки для включения интернета)
- добавил блок try catch на второй экран, с отловом ошибки когда начинают вводить с "."
- поправил момент с "0.", но мне реализация моя не очень понравилась
- разнес корректно swipeToRefresh логику

MrKovalev commented on app/src/main/java/com/currency/converter/base/CurrencyRatesRepository.kt:
```diff
@@ -89,4 +89,6 @@ object CurrencyRatesRepository {
\ No newline at end of file
+}
+
+const val MESSAGE_EXCEPTION_FOR_CALCULATOR = "Введите число, которое не начиначется с  '.'"
```
Хардкод. Локализация на другие языки сломается. Нужно унести это сообщение в ресурсы и использовать оттуда. Например, из презентера ты можешь передавать **R.string.id** твоей строки, а уже на фрагменте делать **context.getString(id)** и достать значение из ресурсов. Тогда у тебя будет работать поддержка разных языков

MrKovalev commented on app/src/main/java/com/currency/converter/base/CustomDialog.kt:
```diff
@@ -1,41 +1,44 @@
+import com.example.converter.databinding.DialogCustomBinding
+
+
 class CustomDialog: DialogFragment() {
```
Давай так и назовем `NetworkAvailabilityDialogFragment`?

MrKovalev commented on app/src/main/java/com/currency/converter/base/CustomDialog.kt:
```diff
@@ -1,41 +1,44 @@
+    override fun onStart() {
+        super.onStart()
+        val width = (resources.displayMetrics.widthPixels * 0.85).toInt()
+        val height = (resources.displayMetrics.heightPixels * 0.40).toInt()
```
Вот эту черную магию можно не делать. Оно сейчас никак не влияет на отображение.
Достаточно будет `dialog?.window?.setLayout(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)` и выглядеть будет также. Сработает так потому что у тебя в верстке итак уже указаны размеры и отступы

MrKovalev commented on app/src/main/java/com/currency/converter/base/CustomDialog.kt:
```diff
@@ -1,41 +1,44 @@
+        super.onStart()
+        val width = (resources.displayMetrics.widthPixels * 0.85).toInt()
+        val height = (resources.displayMetrics.heightPixels * 0.40).toInt()
+        dialog!!.window?.setLayout(width, ViewGroup.LayoutParams.WRAP_CONTENT)
```
от `!!` я бы избавлялся в сторону safe call через `?`

KSVdvlpr replied:
В верхнем примере dialog?.window?.setLayout(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT) уже лишен !!, коммит отдельный заливать не буду.

MrKovalev commented on app/src/main/java/com/currency/converter/base/NetworkRepository.kt:
```diff
@@ -9,7 +9,6 @@ class NetworkRepository (private val application: Application) {
     fun isCheckStatusInternet() : Boolean {
```
Тогда давай назовем `isInternerUnavailable`?

MrKovalev commented on app/src/main/java/com/currency/converter/features/calculator/presenter/CalculatorPresenter.kt:
```diff
@@ -15,70 +16,90 @@ class CalculatorPresenter (private val networkRepository: NetworkRepository) {
+            if (input.isNotEmpty()) {
+                onCurrencyConverted(from, input, to)
+            }
+            if (input == "0." || input == "0") {
```
Ну пока пойдет, только мы по-прежнему можем ввести 00. и все сломается, например. Но пока я бы предложил оставить так и:
- вынести этот кусочек в отдельную функцию `isInputValid()`в котором ты делаешь все проверки введенного инпута и переиспользовать его в `onTextInputChangedOne` и `onTextInputChangedTwo` -> уберем дублирование
- сейчас ты все равно при вводе невалидного инпута посылаешь запрос к серверу на конвертацию, можешь чекнуть в логах. Лишняя работа. Нужно вызывать onCurrencyConverted() только тогда, когда мы уверены что инпут валиден ->  После изменения выше этой проблемы не должно быть

MrKovalev commented on app/src/main/java/com/currency/converter/features/calculator/presenter/CalculatorPresenter.kt:
```diff
@@ -15,70 +16,90 @@ class CalculatorPresenter (private val networkRepository: NetworkRepository) {
+        referenceCurrencyCode: String
+    ) {
+        val value = input.toDouble()
+        onDialogWarning()
```
Ты чекнешь наличие интернета и мб покажешь даже диалог, но все равно попытаешься ниже вызвать запрос, заведомо зная что он не сможет выполнится и будет ошибка сети. Поэтому нужно вызывать репозиторий только когда есть интернет

MrKovalev commented on app/src/main/java/com/currency/converter/features/rate/presenter/RatesPresenter.kt:
```diff
@@ -1,28 +1,33 @@
+    fun onRefreshed(){
+        onSavedCurrencyGated()
+        view?.showRefreshing(refreshing = false)
+        view?.hideProgress()
```
А если интернет вдруг станет медленнее и `onSavedCurrencyGated()` будет выполняться дольше обычного? Тогда состояние загрузки не будет отражать то что ты реально что-то сейчас загружаешь, т.к ты сразу после вызова загрузки остановил крутилку. Нужно завершать показ загрузки только в том случае когда мы реально уже загрузили данные, либо когда свалились с ошибкой

KSVdvlpr replied:
Вот тут не совсем понимаю как переделать. Понимаю, что скорее всего, мне нужно посмотреть как-то на то, завершился метод или нет. А я просто запускаю метод и все.

KSVdvlpr replied:
Я постарался выставить флаги true false, но думаю, что это бесполезно.

MrKovalev replied:
Ага. Надо по-другому. Обсудим сегодня

KSVdvlpr replied:
ок

MrKovalev commented on app/src/main/res/layout/dialog_custom.xml:
```diff
@@ -7,12 +7,20 @@
-            android:layout_margin="15dp"
+            android:id="@+id/imageNoInternetConnection"
+            android:layout_width="30dp"
+            android:layout_height="30dp"
```
По материал гайдлайнам лучше брать с шагом в 4. Т.е 4-8-12-16 и т.д

MrKovalev commented on app/src/main/res/layout/dialog_custom.xml:
```diff
@@ -7,12 +7,20 @@
     <androidx.constraintlayout.widget.ConstraintLayout
         android:layout_width="300dp"
         android:layout_height="140dp"
```
А вот это хак. Лучше стараться избегать таких жесткий указаний размеров. Конкретно в этом случае верстка всего диалога сделана чуток криво. Если покрутить этот диалог на разных девайсах и в разных положениях, с разным текстом у диалога и т.д, то все поедет криво

Накидаю тогда тезисно, попробуй почекать? Не могу над каждой строчкой оставить комменты:
- Зачем дополнительный ConstraintLayout внутри уже существующего рутового ConstraintLayout? Получается излишняя вложенность и сложнее рендерить такой UI. Констрейнт как-раз придумали для того чтобы сделать ее поменьше. Тут можно обойтись только одним ConstraintLayout
- Везде хардкоды размеров вьюх. Например, ширина 140dp и т.д. Нужно стараться верстать так чтобы все было завязано на констрейнты. Т.е правильно выставлять привязки куда и какие вью ты завязываешь. А самим вью делать либо wrap_content(по размеру контента внутри вью), либо 0dp(т.е полностью подчиняться заданным якорям и делать размер под них) либо match_parent и вообще растягивать на всю
- android:background="@color/white" хардкод, сломается темная тема. Нужно использовать атрибут из темы. Тоже самое касается цвета текста, например
- лишние атрибуты вроде layout_gravity, они не работают с констрейнтом. Это скорее в линеаре или релатив лейауте

MrKovalev commented on app/src/main/java/com/currency/converter/features/calculator/presenter/CalculatorPresenter.kt:
```diff
@@ -15,70 +16,90 @@ class CalculatorPresenter (private val networkRepository: NetworkRepository) {
-            onCurrencyConverted(to, input, from)
-        } else {
-            view?.clearTo()
+        try {
```
Не совсем понял зачем тут try/catch. Как будто никакой ошибки возникать не должно. Ты просто должен правильно валидировать ввод пользователя. Тут в пуле писал про это.

Если это кейс когда ты вводишь 0. и у тебя пытается преобразовать это в Double и возникает ошибка, то try/catch тут скорее исправляет следствие, а не причину. Уже произошла ошибка, слишком поздно. Нужно не позволять вызывать преобразование если мы хуйню написали просто. После того как ты сделаешь валидацию - проблема уйдет

KSVdvlpr replied:
Мишань, а вот тут когда начинаешь вводить с "." numberFormatException. Поэтому добавил try catch. Попробовал убрать и провалидировать, не очень получилось у меня, может я конечно криво сделал, но решил пока оставить так.

MrKovalev replied:
Немного по-другому имел ввиду. Обсудим сегодня

KSVdvlpr replied:
ок

MrKovalev reviewed (changes requested):
Вообще, красавчик 👍  Много успел за день. Не напрягайся из-за кучи комментариев. Я просто посмотрел с большей степенью доеба чем делал это раньше. Можешь по комментариям внести правки сюда отдельными коммитами. Если вопросики есть - отвечай мне на мои комментарии

KSVdvlpr committed:
Убрал хардкод,вынес сообщение в ресурсы и использую в презентере с помощью R.string, а во фрагменте context.getString(R.string). Возникла одна сложность передавал в строку '.' и вылетала ошибка, не мог спарсить жти данные.

KSVdvlpr committed:
Изменил naming custom dialog.

KSVdvlpr committed:
Убрал черную магию в NetworkAvailabilityDialogFragment + решил изменить момент с тем, что логика в onStart находится. Сделал функцию и запускаю ее в onViewCreated.

KSVdvlpr committed:
Rename fun isCheckStatusInternet.

KSVdvlpr committed:
Добавил fun isInputValid которая валидирует input по трем значениям : "0." , "0" , ".", плюсом сюда вынес условие if(input.isEmpty()){} .
Добавил дополнительнульную валидацию при вызове onCurrencyConverted input.isNotEmpty() && input != "0." && input != "0", для того, чтобы, когда значение не валидно не создавался новый запрос.
Сделал несколько функций private т.к они используются только здесь.

KSVdvlpr committed:
Добавил дополнительную валидацию, перед вызовом репозитория if (!networkRepository.isInternetUnavailable()) {} с показом диалога view?.showDialog().

KSVdvlpr committed:
Добавил дополнительную валидацию, перед вызовом репозитория if (!networkRepository.isInternetUnavailable()) {} с показом диалога view?.showDialog(). так же как и на втором экране, чтобы дополнительно не идти за запросом, когда нет интернета.
+ переделал showRefreshing, но скорее всего неправильно. Проверил дебагерром, методы вызывается, но не завершается.

KSVdvlpr committed:
Прошелся по всей верстке: постарался убрать все хардкоды, разобрался с цветами и темной темой, убрал вложенный constraintLayout, gravity почистил.

KSVdvlpr committed:
последние добавления:
- правильная валидация ввода на 2 экране
- корректная остановка swipeToRefresh
- try catch убран
